# Bigrams

In [1]:
import pandas as pd
import string
import nltk
from nltk.corpus import stopwords
import seaborn as sns
from nltk.tokenize import word_tokenize 
from nltk.collocations import *
import nltk
from nltk.stem import WordNetLemmatizer
import seaborn as sns

In [2]:
papers = pd.read_csv("tabela_1_ed_sem_2019.csv", index_col=[0]) #dataframe derived from medline file
papers.head(5)

Journal  \
0  International journal of systematic and evolut...   
1  International journal of systematic and evolut...   
4  International journal of systematic and evolut...   
5  International journal of systematic and evolut...   
6  International journal of systematic and evolut...   

                                               Title  \
0  Leucothrix arctica sp. nov., isolated from Arc...   
1  Hyphobacterium indicum sp. nov., isolated from...   
4  Sphingopyxis lindanitolerans sp. nov. strain W...   
5  Mesosutterella multiformis gen. nov., sp. nov....   
6  Acinetobacter sichuanensis sp. nov., recovered...   

                                             Authors  Year  \
0  ['Baek Kiwoon' 'Choi Ahyoung' 'Lee Yung Mi' 'L...  2018   
1  ['Ruan Chu-Jin' 'Zheng Xiao-Wei' 'Wang Jian' '...  2018   
4  ['Kaminski Michal A' 'Sobczak Adam' 'Spolnik G...  2018   
5  ['Sakamoto Mitsuo' 'Ikeyama Nao' 'Kunihiro Tad...  2018   
6  ['Qin Jiayuan' 'Hu Yiyi' 'Feng Yu' 'Lv Xiaoju'...  2018   

                                            Abstract  
0  A Gram-stain-negative, non-motile, oxidase- an...  
1  A novel aerobic, Gram-stain-negative bacterium...  
4  An aerobic, Gram-stain-negative, rod-shaped, n...  
5  Two novel, obligately anaerobic, Gram-stain-ne...  
6  A novel Acinetobacter strain, WCHAc060041(T), ...

In [3]:
abstracts_full = papers[["Abstract", "Year"]]

In [4]:
stop = stopwords.words('english')

In [5]:
def freq_bigram_by_year(year):
    abstracts_year = abstracts_full[abstracts_full["Year"] == year]["Abstract"]
    
    abstracts_year_cat = abstracts_year.str.cat(sep=" ")
    
    punct = '!"#$%&\'()*+,./:;<=>?@[\\]^_`{}~'   # `|` is not present here
    transtab = str.maketrans(dict.fromkeys(punct, ''))
    abstracts_year_no_punct = abstracts_year_cat.translate(transtab)
    
    stop_words = stopwords.words('english')
    word_tokens_year = word_tokenize(abstracts_year_no_punct.lower()) 
    
    filtered_sentence_year = [] 
    for w in word_tokens_year: 
        if w not in stop_words: 
            filtered_sentence_year.append(w)
    
    wordnet_lemmatizer = WordNetLemmatizer()
    lemmatized_year = []
    
    
    for w in filtered_sentence_year:
        lemmatized_w = wordnet_lemmatizer.lemmatize(w)
        lemmatized_year.append(lemmatized_w)
    
    #bigrams
    bigramFinder_year = nltk.collocations.BigramCollocationFinder.from_words(lemmatized_year)
    bigram_freq_year = bigramFinder_year.ngram_fd.items()
    bigramFreqTable_year = pd.DataFrame(list(bigram_freq_year), columns=['bigram','freq']).sort_values(by='freq', ascending=False)
    bigramFreqTable_year.to_csv("{}_bigram_freq.csv".format(year))
 

In [6]:
for i in range(2001,2019):
    freq_bigram_by_year(i)

In [7]:
#frequency from all
abstracts_all = abstracts_full["Abstract"]
abstracts_cat = abstracts_all.str.cat(sep=" ")

punct = '!"#$%&\'()*+,./:;<=>?@[\\]^_`{}~'   # `|` is not present here
transtab = str.maketrans(dict.fromkeys(punct, ''))
abstracts_all_no_punct = abstracts_cat.translate(transtab)

stop_words = stopwords.words('english')
word_tokens = word_tokenize(abstracts_all_no_punct.lower()) 
  
filtered_sentence_all = [] 
  
for w in word_tokens: 
    if w not in stop_words: 
        filtered_sentence_all.append(w)    

wordnet_lemmatizer = WordNetLemmatizer()
lemmatized_all = []

for w in filtered_sentence_all:
    lemmatized_w = wordnet_lemmatizer.lemmatize(w)
    lemmatized_all.append(lemmatized_w)
        
bigramFinder_all = nltk.collocations.BigramCollocationFinder.from_words(lemmatized_all)
bigram_freq_all = bigramFinder_all.ngram_fd.items()
bigramFreqTable_all = pd.DataFrame(list(bigram_freq_all), columns=['bigram','freq']).sort_values(by='freq', ascending=False)

bigramFreqTable_all

bigram   freq
28                             (type, strain)  13821
16                                 (16, rrna)  13082
17                               (rrna, gene)  12456
18                           (gene, sequence)  11829
129                                 (sp, nov)  10471
78                              (fatty, acid)   8252
123                           (novel, specie)   8084
124                           (specie, genus)   6813
40                     (sequence, similarity)   6794
130                           (nov, proposed)   6578
69                                (degree, c)   6487
59                              (gc, content)   6397
131                          (proposed, type)   5644
185                  (phylogenetic, analysis)   5275
58                                  (dna, gc)   4642
104                            (polar, lipid)   4207
187                               (based, 16)   4007
26                         (closely, related)   3723
122                       (represents, novel)   3287
186                         (analysis, based)   3163
77                             (major, fatty)   3112
340                  (dna-dna, hybridization)   3010
19                       (sequence, analysis)   2715
678                           (member, genus)   2498
70                                    (c, ph)   2497
1298                              (nov, type)   2440
942                    (dna-dna, relatedness)   2434
620                         (cellular, fatty)   2371
80                          (summed, feature)   2224
1520                             (iso-c15, 0)   2173
...                                       ...    ...
170657                            (nnt, 9687)      1
170670                     (g, xinjiangensis)      1
170654                         (tp2-8t, 9706)      1
170652                     (similarity, 9706)      1
170651                       (j2, similarity)      1
170649            (relative, gracilibacillus)      1
170648             (gracilibacillus, closest)      1
170647                       (sg103t, member)      1
170646                    (suggested, sg103t)      1
170669                             (9687, 16)      1
170671                                (j2, g)      1
170692      (plant-growth-promoting, ability)      1
170684                          (sg103t, jcm)      1
170691   (evaluation, plant-growth-promoting)      1
170690                     (winteri, natural)      1
170689  (phosphate-mobilizing, gram-negative)      1
170688         (14683t, phosphate-mobilizing)      1
170687                        (ncimb, 14683t)      1
170686                        (17253t, ncimb)      1
170685                          (jcm, 17253t)      1
170683                     (alcaliphilus, sp)      1
170673                            (tp2-8t, g)      1
170682        (gracilibacillus, alcaliphilus)      1
170681                        (sg103t, merit)      1
170679                     (3-2, relatedness)      1
170678                             (4-9, 3-2)      1
170677                             (2-4, 4-9)      1
170676                        (revealed, 2-4)      1
170675                        (nnt, revealed)      1
408879              (antitubercular, therapy)      1

[408880 rows x 2 columns]

In [8]:
n_year = pd.DataFrame(abstracts_full["Year"].value_counts()).reset_index()
n_year


index  Year
0    2015   903
1    2017   867
2    2013   853
3    2016   775
4    2014   757
5    2018   647
6    2012   560
7    2010   534
8    2011   534
9    2009   529
10   2007   502
11   2008   486
12   2006   416
13   2005   376
14   2004   330
15   2003   269
16   2002   234
17   2001   179

In [9]:
#itens per year
n_year.columns = ["Year", "Frequency"]
n_year = n_year.sort_values(by="Year")
n_year_ed = n_year.set_index("Year")
n_year_transposed = n_year_ed.transpose()
n_year_transposed.columns = ["2001", "2002", "2003", "2004", "2005", "2006", "2007", "2008", "2009", "2010", "2011", "2012", "2013", "2014", "2015", "2016", "2017", "2018"]
n_year_transposed

2001  2002  2003  2004  2005  2006  2007  2008  2009  2010  2011  \
Frequency   179   234   269   330   376   416   502   486   529   534   534   

           2012  2013  2014  2015  2016  2017  2018  
Frequency   560   853   757   903   775   867   647

In [10]:
bigramFreqTable_2001 = pd.read_table('2001_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2002 = pd.read_table('2002_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2003 = pd.read_table('2003_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2004 = pd.read_table('2004_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2005 = pd.read_table('2005_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2006 = pd.read_table('2006_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2007 = pd.read_table('2007_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2008 = pd.read_table('2008_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2009 = pd.read_table('2009_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2010 = pd.read_table('2010_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2011 = pd.read_table('2011_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2012 = pd.read_table('2012_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2013 = pd.read_table('2013_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2014 = pd.read_table('2014_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2015 = pd.read_table('2015_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2016 = pd.read_table('2016_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2017 = pd.read_table('2017_bigram_freq.csv', sep=',', index_col=0)
bigramFreqTable_2018 = pd.read_table('2018_bigram_freq.csv', sep=',', index_col=0)

dfs = [bigramFreqTable_2001, bigramFreqTable_2002, bigramFreqTable_2003, bigramFreqTable_2004, bigramFreqTable_2005, bigramFreqTable_2006, bigramFreqTable_2007, bigramFreqTable_2008, bigramFreqTable_2009, bigramFreqTable_2010, bigramFreqTable_2011, bigramFreqTable_2012, bigramFreqTable_2013, bigramFreqTable_2014, bigramFreqTable_2015, bigramFreqTable_2016, bigramFreqTable_2017, bigramFreqTable_2018]
from functools import reduce

In [11]:
df_final = reduce(lambda left,right: pd.merge(left,right,on='bigram'), dfs)
df_final.columns = ["bigram", "2001", "2002", "2003", "2004", "2005", "2006", "2007", "2008", "2009", "2010", "2011", "2012", "2013", "2014", "2015", "2016", "2017", "2018" ]

In [12]:
df_final = df_final.set_index("bigram")

In [13]:
df_final.head(3)

2001  2002  2003  2004  2005  2006  2007  2008  2009  \
bigram                                                                     
('type', 'strain')   195   278   283   401   481   573   703   743   830   
('sp', 'nov')        175   235   253   382   426   482   576   545   614   
('16', 'rrna')        93   135   153   324   495   598   730   719   761   

                    2010  2011  2012  2013  2014  2015  2016  2017  2018  
bigram                                                                    
('type', 'strain')   796   744   813  1280  1182  1336  1100  1208   875  
('sp', 'nov')        584   575   614   896   800   939   815   900   660  
('16', 'rrna')       785   799   795  1214  1097  1246  1090  1187   861

In [14]:
relative = df_final.div(n_year_transposed.iloc[0])

In [15]:
transposed_relative = relative.T
transposed_relative

bigram  ('type', 'strain')  ('sp', 'nov')  ('16', 'rrna')  ('new', 'specie')  \
2001              1.089385       0.977654        0.519553           0.508380   
2002              1.188034       1.004274        0.576923           0.209402   
2003              1.052045       0.940520        0.568773           0.189591   
2004              1.215152       1.157576        0.981818           0.127273   
2005              1.279255       1.132979        1.316489           0.106383   
2006              1.377404       1.158654        1.437500           0.069712   
2007              1.400398       1.147410        1.454183           0.053785   
2008              1.528807       1.121399        1.479424           0.028807   
2009              1.568998       1.160681        1.438563           0.034026   
2010              1.490637       1.093633        1.470037           0.041199   
2011              1.393258       1.076779        1.496255           0.046816   
2012              1.451786       1.096429        1.419643           0.046429   
2013              1.500586       1.050410        1.423212           0.029308   
2014              1.561427       1.056803        1.449141           0.033025   
2015              1.479513       1.039867        1.379845           0.036545   
2016              1.419355       1.051613        1.406452           0.043871   
2017              1.393310       1.038062        1.369089           0.043829   
2018              1.352396       1.020093        1.330757           0.075734   

bigram  ('degree', 'c')  ('fatty', 'acid')  ('gc', 'content')  \
2001           0.424581           0.407821           0.391061   
2002           0.529915           0.538462           0.384615   
2003           0.602230           0.579926           0.237918   
2004           0.412121           0.587879           0.375758   
2005           0.465426           0.712766           0.494681   
2006           0.473558           0.694712           0.533654   
2007           0.527888           0.846614           0.643426   
2008           0.582305           0.775720           0.635802   
2009           0.578450           0.780718           0.631380   
2010           0.565543           0.855805           0.653558   
2011           0.617978           0.883895           0.677903   
2012           0.651786           0.914286           0.694643   
2013           0.706917           0.917937           0.715123   
2014           0.793923           0.945839           0.760898   
2015           0.740864           0.940199           0.720930   
2016           0.776774           0.934194           0.740645   
2017           0.843137           0.944637           0.795848   
2018           0.865533           0.925811           0.732612   

bigram  ('rrna', 'gene')  ('nov', 'type')  ('dna-dna', 'hybridization')  \
2001            0.363128         0.363128                      0.357542   
2002            0.457265         0.363248                      0.354701   
2003            0.501859         0.304833                      0.308550   
2004            0.909091         0.351515                      0.324242   
2005            1.263298         0.369681                      0.356383   
2006            1.389423         0.382212                      0.336538   
2007            1.392430         0.256972                      0.294821   
2008            1.425926         0.288066                      0.327160   
2009            1.400756         0.323251                      0.332703   
2010            1.406367         0.314607                      0.277154   
2011            1.443820         0.217228                      0.303371   
2012            1.376786         0.246429                      0.258929   
2013            1.350528         0.202814                      0.322392   
2014            1.377807         0.269485                      0.266843   
2015            1.312292         0.198228                      0.314507   
2016            1.343226         0.184516  

In [16]:
new_specie = transposed_relative["('sp', 'nov')"]
new_specie.index.name = 'year'
new_specie = new_specie.reset_index()
new_specie.columns = ["year", "freq"]
new_specie.columns

Index(['year', 'freq'], dtype='object')

In [17]:
new_genus = transposed_relative["('gen', 'nov')"]
new_genus.index.name = 'year'
new_genus = new_genus.reset_index()
new_genus.columns = ["year", "freq"]
new_genus.columns

Index(['year', 'freq'], dtype='object')

In [18]:
s16_rrna = transposed_relative["('16', 'rrna')"]
s16_rrna.index.name = 'year'
s16_rrna = s16_rrna.reset_index()
s16_rrna.columns = ["year", "freq"]
s16_rrna.columns

Index(['year', 'freq'], dtype='object')

In [19]:
fatty_acid = transposed_relative["('fatty', 'acid')"]
fatty_acid.index.name = 'year'
fatty_acid = fatty_acid.reset_index()
fatty_acid.columns = ["year", "freq"]
fatty_acid.columns

Index(['year', 'freq'], dtype='object')

In [20]:
gc_content = transposed_relative["('gc', 'content')"]
gc_content.index.name = 'year'
gc_content = gc_content.reset_index()
gc_content.columns = ["year", "freq"]
gc_content.columns

Index(['year', 'freq'], dtype='object')

In [21]:
phylogenetic_analysis = transposed_relative["('phylogenetic', 'analysis')"]
phylogenetic_analysis.index.name = 'year'
phylogenetic_analysis = phylogenetic_analysis.reset_index()
phylogenetic_analysis.columns = ["year", "freq"]
phylogenetic_analysis.columns

Index(['year', 'freq'], dtype='object')

In [22]:
polar_lipid = transposed_relative["('polar', 'lipid')"]
polar_lipid.index.name = 'year'
polar_lipid = polar_lipid.reset_index()
polar_lipid.columns = ["year", "freq"]
polar_lipid.columns

Index(['year', 'freq'], dtype='object')

In [23]:
dna_hybridization = transposed_relative["('dna-dna', 'hybridization')"]
dna_hybridization.index.name = 'year'
dna_hybridization = dna_hybridization.reset_index()
dna_hybridization.columns = ["year", "freq"]
dna_hybridization.columns

Index(['year', 'freq'], dtype='object')

In [24]:
respiratory_quinone = transposed_relative["('respiratory', 'quinone')"]
respiratory_quinone.index.name = 'year'
respiratory_quinone = respiratory_quinone.reset_index()
respiratory_quinone.columns = ["year", "freq"]
respiratory_quinone.columns

Index(['year', 'freq'], dtype='object')

# Figure 8

import matplotlib.pyplot as plt
from matplotlib import rcParams

fig, ax = plt.subplots(figsize=(12,8))

for axis in ['top','bottom','left','right']:
    ax.spines[axis].set_linewidth(1.5)

ax.xaxis.set_tick_params(width=1.5, labelsize=14)
ax.yaxis.set_tick_params(width=1.5, labelsize=14)
#pal = sns.color_palette("Set1")
rcParams['font.family'] = 'sans-serif'
rcParams['font.sans-serif'] = ['Arial']

sns.pointplot(x="year", y="freq", data=s16_rrna, color="#f77189")
fig.text(x = 0.91, y = 0.76, s = "16 rrna", family='sans-serif', fontsize=14, weight = "bold", color="#f77189")
sns.pointplot(x="year", y="freq", data=fatty_acid, color="#bb9832")
fig.text(x = 0.91, y = 0.57, s = "fatty acid", family='sans-serif', fontsize=14, weight = "bold", color="#bb9832")
sns.pointplot(x="year", y="freq", data=gc_content, color="#50b131")
fig.text(x = 0.91, y = 0.48, s = "gc content", family='sans-serif', fontsize=14, weight = "bold", color="#50b131")
sns.pointplot(x="year", y="freq", data=phylogenetic_analysis, color="#36ada4")
fig.text(x = 0.91, y = 0.42, s = "phylogenetic analysis", family='sans-serif', fontsize=14, weight = "bold", color="#36ada4")
sns.pointplot(x="year", y="freq", data=polar_lipid, color="#3ba3ec")
fig.text(x = 0.91, y = 0.50, s = "polar lipid", family='sans-serif', fontsize=14, weight = "bold", color="#3ba3ec")
sns.pointplot(x="year", y="freq", data=dna_hybridization, color="#e866f4")
fig.text(x = 0.91, y = 0.32, s = "dna-dna hybridization", family='sans-serif', fontsize=14, weight = "bold", color="#e866f4")
sns.pointplot(x="year", y="freq", data=respiratory_quinone, color="#a48cf4")
fig.text(x = 0.91, y = 0.30, s = "respiratory quinone", weight = "bold", family='sans-serif', fontsize=14, color="#a48cf4")

fig.savefig('bigram.svg', format='svg', dpi=600)